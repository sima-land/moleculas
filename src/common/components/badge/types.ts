import type { AnchorHTMLAttributes, CSSProperties } from 'react';

export interface BadgeField {
  type: 'text' | 'timer' | 'svg-url';
  value: string;
}

export interface BadgeStyle extends CSSProperties {
  '--badge-color'?: string;
  '--badge-text-color'?: string;
  '--badge-border-color'?: string;
}

export interface BadgeProps extends Omit<AnchorHTMLAttributes<HTMLAnchorElement>, 'color'> {
  /** Основной цвет, используется как цвет обводки, иконок и текста. */
  color?: string;

  /** Как применится цвет. */
  coloring?: 'fill' | 'outline' | 'unset';

  /** Содержимое шильдика. */
  fields?: Array<BadgeField>;

  /** Форма скруглений. */
  shape?: 'round' | 'pill' | 'unset';

  /** Нужен ли эффект при наведении. */
  withHoverEffect?: boolean;

  /** Стили. */
  style?: BadgeStyle;

  /** Идентификатор для систем автоматизированного тестирования. */
  'data-testid'?: string;
}
