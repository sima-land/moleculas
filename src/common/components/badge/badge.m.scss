@use 'pkg:@sima-land/ui-nucleons/utils.scss';

$badge-size: 20px;
$icon-size: 16px;

.root {
  @include utils.reset-button;
}

.root {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  height: $badge-size;
  font-weight: bold;
  font-size: 12px;
  line-height: 16px;
  max-width: 100%;
  min-width: $badge-size;
  text-decoration: none;
}

.padding-x-default {
  padding-left: 6px;
  padding-right: 6px;
  &.coloring-outline {
    // ВАЖНО: учитываем border
    padding-left: 5px;
    padding-right: 5px;
  }
}

.padding-x-pill {
  padding-left: 8px;
  padding-right: 8px;
  &.coloring-outline {
    // ВАЖНО: учитываем border
    padding-left: 7px;
    padding-right: 7px;
  }
}

.shape-round {
  border-radius: 4px;
}

.shape-pill {
  border-radius: 99999px;
}

.with-hover-effect:hover {
  position: relative;
  cursor: pointer;
  &::after {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background: #fff;
    opacity: 0.16;
  }
}

.coloring-outline {
  color: var(--badge-color, #aaa);

  // ВАЖНО: используем border т.к. на Safari box-shadow пропадает при уменьшении масштаба
  border: 1px solid var(--badge-border-color, var(--badge-color, #aaa));
}

.coloring-fill {
  background: var(--badge-color, #aaa);
  color: var(--badge-text-color, #fff);
}

.icon-only {
  width: $badge-size;
}

// содержимое
.content {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.icon {
  display: block;
  width: $icon-size;
  height: $icon-size;
  line-height: $icon-size;
}

.icon + .content,
.content + .icon {
  margin-left: 4px;
}
