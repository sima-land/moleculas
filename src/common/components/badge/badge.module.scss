$badge-size: 20px;
$icon-size: 14px;

.root {
  --badge-color: #aaa;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  height: $badge-size;
  color: var(--badge-color);

  // используем border т.к. на Safari box-shadow пропадает при уменьшении масштаба
  border: 1px solid var(--badge-color);
  padding: 0 5px; // учитываем border
  font-weight: bold;
  font-size: 12px;
  line-height: 16px;
  max-width: 100%;
  min-width: $badge-size;
  text-decoration: none;
  &.interactive:hover {
    cursor: pointer;
    opacity: 0.72;
  }
  &.icon-only {
    width: $badge-size;
    padding: 0;
  }
  .content {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
  .icon {
    display: inline-block;
    width: $icon-size;
    height: $icon-size;
    line-height: $icon-size;
  }
}
