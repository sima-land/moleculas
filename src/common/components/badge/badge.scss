$badge-size: 20px;
$icon-size: 14px;

.root {
  --badge-color: #aaa;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  height: $badge-size;
  color: var(--badge-color);
  box-shadow: inset 0 0 0 1px var(--badge-color);
  padding: 0 6px;
  font-weight: bold;
  font-size: 12px;
  line-height: 16px;
  max-width: 100%;
  min-width: $badge-size;
  text-decoration: none;
  &.interactive:hover {
    cursor: pointer;
    opacity: .72;
  }
  &.icon-only {
    width: $badge-size;
    padding: 0;
  }
  .content {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
  .svg-wrapper {
    width: $icon-size;
    height: $icon-size;
    svg {
      fill: var(--badge-color);
      display: inline-block;
      width: $icon-size;
      height: $icon-size;
      line-height: $icon-size;

      // чтобы перекрыть inline-свойство fill на path, circle и тд
      * {
        fill: inherit;
      }
    }
  }
}
