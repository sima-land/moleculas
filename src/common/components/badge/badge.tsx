import { ReactNode, forwardRef } from 'react';
import type { BadgeProps, BadgeStyle } from './types';
import { Timer } from '@sima-land/ui-nucleons/timer';
import classnames from 'classnames/bind';
import styles from './badge.module.scss';

const cx = classnames.bind(styles);

/**
 * Шильдик.
 * @param props Свойства.
 * @return Элемент.
 */
export const Badge = forwardRef<HTMLAnchorElement, BadgeProps>(function Badge(
  {
    className,
    color,
    coloring = 'outline',
    fields,
    href,
    style,
    shape = 'unset',
    withHoverEffect = Boolean(href),
    'data-testid': testId = 'badge',
    ...restProps
  },
  ref,
) {
  const iconOnly = fields.length === 1 && fields[0].type === 'svg-url';

  const rootClassName = cx(
    'root',
    iconOnly && 'icon-only',
    withHoverEffect && 'with-hover-effect',
    shape === 'round' && 'shape-round',
    shape === 'pill' && 'shape-pill',
    shape !== 'pill' && !iconOnly && 'padding-x-default',
    shape === 'pill' && !iconOnly && 'padding-x-pill',
    coloring === 'outline' && 'coloring-outline',
    coloring === 'fill' && 'coloring-fill',
    className,
  );

  const rootStyle: BadgeStyle = {
    '--badge-color': color,
    ...style,
  };

  return (
    <a
      ref={ref}
      {...restProps}
      data-testid={testId}
      href={href}
      style={rootStyle}
      className={rootClassName}
      children={
        iconOnly ? (
          <img className={cx('icon')} src={fields[0].value} alt='' />
        ) : (
          // вложенный span нужен для того чтобы объединить `display: inline-flex` и `text-overflow: ellipsis`
          <span className={cx('content')}>
            {fields.reduce<ReactNode[]>((acc, item, i) => {
              let result = null;

              // на данный момент сочетание иконки с другими элементами не предусмотрено макетами - игнорируем
              switch (item.type) {
                case 'text':
                  result = item.value;
                  break;
                case 'timer':
                  result = (
                    <Timer key={i} date={item.value} timeout={1000 * 60} format={formatDistance} />
                  );
                  break;
              }

              acc.push(result);

              i < fields.length - 1 && acc.push(' ');

              return acc;
            }, [])}
          </span>
        )
      }
    />
  );
});

/**
 * Форматирует оставшееся время.
 * @param distance Оставшееся время.
 * @return Отформатированное время.
 */
const formatDistance = ({
  days,
  hours,
  minutes,
}: {
  days: number;
  hours: number;
  minutes: number;
}) => [toTimePart(days), ...[hours % 24, minutes % 60].map(toTimePart)].join(':');

/**
 * Форматирует число добавляя нули при необходимости.
 * @param n Число.
 * @return Отформатированное число.
 */
const toTimePart = (n: number) => `${n}`.padStart(2, '0');
