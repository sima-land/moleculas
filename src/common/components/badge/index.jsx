import React from 'react';
import classnames from 'classnames/bind';
import { Timer } from '@dev-dep/ui-nucleons/timer';
import styles from './badge.scss';

const cx = classnames.bind(styles);

/**
 * Шильдик.
 * @param {Object} props Свойства.
 * @param {string} [props.className] Класс.
 * @param {string} props.color Цвет.
 * @param {Array<{ type: 'text' | 'timer' | 'svg-url', value: string }>} props.fields Содержимое.
 * @param {string} props.href Ссылка.
 * @param {Object} [props.style] Стили.
 * @return {ReactElement} Компонент.
 */
export const Badge = ({
  className,
  color,
  fields,
  href,
  style,
  ...restProps
}) => {
  const Element = href ? 'a' : 'span';
  const iconOnly = fields.length === 1 && fields[0].type === 'svg-url';

  return (
    <Element
      {...restProps}
      href={href}
      style={{ ...style, '--badge-color': color }}
      className={cx(
        'root',
        iconOnly && 'icon-only',
        href && 'interactive',
        className
      )}
      children={
        iconOnly
          ? (
            <img
              className={cx('icon')}
              src={fields[0].value}
            />
          ) : (

            // вложенный span нужен для того чтобы объединить `display: inline-flex` и `text-overflow: ellipsis`
            <span className={cx('content')}>
              {fields.reduce((acc, item, i) => {
                let result = null;

                // на данный момент сочетание иконки с другими элементами не предусмотрено макетами - игнорируем
                switch (item.type) {
                  case 'text':
                    result = item.value;
                    break;
                  case 'timer':
                    result = (
                      <Timer
                        key={i}
                        date={item.value}
                        timeout={1000 * 60}
                        format={formatDistance}
                      />
                    );
                    break;
                }

                acc.push(result);
                i < fields.length - 1 && acc.push(' ');

                return acc;
              }, [])}
            </span>
          )
      }
    />
  );
};

/**
 * Форматирует оставшееся время.
 * @param {{ days, hours, minutes }} distance Оставшееся время.
 * @return {string} Отформатированное время.
 */
const formatDistance = ({ days, hours, minutes }) => [
  toTimePart(days),
  ...[hours % 24, minutes % 60].map(toTimePart),
].join(':');

/**
 * Форматирует число добавляя нули при необходимости.
 * @param {number} n Число.
 * @return {string} Отформатированное число.
 */
const toTimePart = n => `${n}`.padStart(2, '0');
