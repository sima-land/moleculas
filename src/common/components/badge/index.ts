export { type BadgeProps, type BadgeField, Badge } from './badge';
