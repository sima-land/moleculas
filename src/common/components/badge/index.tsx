import React from 'react';
import classnames from 'classnames/bind';
import { Timer } from '@sima-land/ui-nucleons/timer';
import styles from './badge.module.scss';

export interface BadgeField {
  type: 'text' | 'timer' | 'svg-url'
  value: string
}

export interface BadgeProps extends React.HTMLAttributes<HTMLSpanElement> {

  /** Основной цвет, используется как цвет обводки, иконок и текста. */
  color?: string

  /** Содержимое шильдика. */
  fields: Array<BadgeField>

  /** Ссылка, переход по которой будет произведён по нажатию на шильдик. */
  href?: string

  /** Идентификатор для систем автоматизированного тестирования. */
  'data-testid'?: string
}

const cx = classnames.bind(styles);

/**
 * Шильдик.
 * @param props Свойства.
 * @param props.color Цвет.
 * @param props.fields Содержимое.
 * @param props.href Ссылка.
 * @return Элемент.
 */
export const Badge = ({
  className,
  color,
  fields,
  href,
  style,
  'data-testid': testId = 'badge',
  ...restProps
}: BadgeProps) => {
  const Element = href ? 'a' : 'span';
  const iconOnly = fields.length === 1 && fields[0].type === 'svg-url';

  return (
    <Element
      {...restProps}
      data-testid={testId}
      href={href}
      style={{ ...style, '--badge-color': color } as any}
      className={cx(
        'root',
        iconOnly && 'icon-only',
        href && 'interactive',
        className
      )}
      children={
        iconOnly
          ? (
            <img
              className={cx('icon')}
              src={fields[0].value}
            />
          ) : (

            // вложенный span нужен для того чтобы объединить `display: inline-flex` и `text-overflow: ellipsis`
            <span className={cx('content')}>
              {fields.reduce<React.ReactNode[]>((acc, item, i) => {
                let result = null;

                // на данный момент сочетание иконки с другими элементами не предусмотрено макетами - игнорируем
                switch (item.type) {
                  case 'text':
                    result = item.value;
                    break;
                  case 'timer':
                    result = (
                      <Timer
                        key={i}
                        date={item.value}
                        timeout={1000 * 60}
                        format={formatDistance}
                      />
                    );
                    break;
                }

                acc.push(result);
                i < fields.length - 1 && acc.push(' ');

                return acc;
              }, [])}
            </span>
          )
      }
    />
  );
};

/**
 * Форматирует оставшееся время.
 * @param distance Оставшееся время.
 * @return Отформатированное время.
 */
const formatDistance = ({
  days,
  hours,
  minutes,
}: {
  days: number
  hours: number
  minutes: number
}) => [
  toTimePart(days),
  ...[hours % 24, minutes % 60].map(toTimePart),
].join(':');

/**
 * Форматирует число добавляя нули при необходимости.
 * @param n Число.
 * @return Отформатированное число.
 */
const toTimePart = (n: number) => `${n}`.padStart(2, '0');
