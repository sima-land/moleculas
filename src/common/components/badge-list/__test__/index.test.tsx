import React from 'react';
import { BadgeList } from '..';
import { Badge, BadgeProps } from '../../badge';
import { render } from '@testing-library/react';

const badges: BadgeProps[] = [
  { color: '#b52ea8', fields: [{ type: 'text', value: '-56%' }] },
  { color: '#ff7200', fields: [{ type: 'text', value: '3 по цене 2' }] },
  { color: '#2962ff', fields: [{ type: 'text', value: 'Товар месяца' }] },
  { color: '#00b8d4', fields: [{ type: 'text', value: 'Акция' }] },
];

describe('<BadgeList />', () => {
  it('renders correctly', () => {
    const { container, getAllByTestId } = render(
      <BadgeList className='test'>
        {badges.map((item, index) => (
          <Badge key={index} {...item} />
        ))}
      </BadgeList>,
    );

    expect(container).toMatchSnapshot();
    expect(getAllByTestId('badge')).toHaveLength(4);
  });
});
