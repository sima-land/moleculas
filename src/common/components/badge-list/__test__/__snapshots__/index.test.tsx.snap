// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BadgeList /> renders correctly 1`] = `
<div>
  <div
    class="root test"
  >
    <span
      class="root item"
      data-testid="badge"
      style="--badge-color: #b52ea8;"
    >
      <span
        class="content"
      >
        -56%
      </span>
    </span>
    <span
      class="root item"
      data-testid="badge"
      style="--badge-color: #ff7200;"
    >
      <span
        class="content"
      >
        3 по цене 2
      </span>
    </span>
    <span
      class="root item"
      data-testid="badge"
      style="--badge-color: #2962ff;"
    >
      <span
        class="content"
      >
        Товар месяца
      </span>
    </span>
    <span
      class="root item"
      data-testid="badge"
      style="--badge-color: #00b8d4;"
    >
      <span
        class="content"
      >
        Акция
      </span>
    </span>
  </div>
</div>
`;

exports[`<BadgeList /> with props renders correctly 1`] = `
<BadgeList
  className="test"
  items={
    Array [
      Object {
        "color": "#64a0eb",
        "fields": Array [
          Object {
            "type": "text",
            "value": "-10%",
          },
        ],
        "href": "/sale/?is_label=1",
      },
    ]
  }
>
  <div
    className="root test"
  >
    <Badge
      className="item"
      color="#64a0eb"
      fields={
        Array [
          Object {
            "type": "text",
            "value": "-10%",
          },
        ]
      }
      href="/sale/?is_label=1"
      key="0"
    >
      <a
        className="root interactive item"
        data-testid="badge"
        href="/sale/?is_label=1"
        style={
          Object {
            "--badge-color": "#64a0eb",
          }
        }
      >
        <span
          className="content"
        >
          -10%
        </span>
      </a>
    </Badge>
  </div>
</BadgeList>
`;

exports[`<BadgeList /> without props renders correctly 1`] = `
<BadgeList>
  <div
    className="root"
  />
</BadgeList>
`;
